pkg/server/policy/policy_test.go

```diff
@@ -0,0 +1,262 @@
+					"some_field": "abc",
+				},
+			},
+			expectResult: policy.Result{
+				Allow:             true,
+				AllowIfAdmin:      false,
+				AllowIfLocal:      false,
+				AllowIfDownstream: false,
+				AllowIfAgent:      false,
+			},
+		},
+		{
+			name:     "test policy with input full method",
+			rego:     condCheckRego("input.full_method == \"some_method\""),
+			jsonData: "{}",
+			input: policy.Input{
+				Caller:     "some_caller",
+				FullMethod: "some_method",
+				Req: map[string]interface{}{
+					"some_field": "abc",
+				},
+			},
+			expectResult: policy.Result{
+				Allow:             true,
+				AllowIfAdmin:      false,
+				AllowIfLocal:      false,
+				AllowIfDownstream: false,
+				AllowIfAgent:      false,
+			},
+		},
+		{
+			name:     "test policy with req field comparison",
+			rego:     condCheckRego("input.req.some_field == \"abc\""),
+			jsonData: "{}",
+			input: policy.Input{
+				Caller:     "some_caller",
+				FullMethod: "some_method",
+				Req: map[string]interface{}{
+					"some_field": "abc",
+				},
+			},
+			expectResult: policy.Result{
+				Allow:             true,
+				AllowIfAdmin:      false,
+				AllowIfLocal:      false,
+				AllowIfDownstream: false,
+				AllowIfAgent:      false,
+			},
+		},
+		{
+			name:     "test policy with req nested field comparison",
+			rego:     condCheckRego("input.req.nested.field == \"def\""),
+			jsonData: "{}",
+			input: policy.Input{
+				Caller:     "some_caller",
+				FullMethod: "some_method",
+				Req: map[string]interface{}{
+					"some_field": "abc",
+					"nested": map[string]interface{}{
+						"field": "def",
+					},
+				},
+			},
+			expectResult: policy.Result{
+				Allow:             true,
+				AllowIfAdmin:      false,
+				AllowIfLocal:      false,
+				AllowIfDownstream: false,
+				AllowIfAgent:      false,
+			},
+		},
+		{
+			name:     "test policy with data bindings",
+			rego:     condCheckRego("input.req.some_field == data.datafield1"),
+			jsonData: `{ "datafield1":"data1"}`,
+			input: policy.Input{
+				Caller:     "some_caller",
+				FullMethod: "some_method",
+				Req: map[string]interface{}{
+					"some_field": "data1",
+				},
+			},
+			expectResult: policy.Result{
+				Allow:             true,
+				AllowIfAdmin:      false,
+				AllowIfLocal:      false,
+				AllowIfDownstream: false,
+				AllowIfAgent:      false,
+			},
+		},
+
+		/*
+		   Test field condition
+		   Test nested field conditions
+		   Test data JSON
+		   Test if all admin thing
+		*/
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			var json map[string]interface{}
+			err := util.UnmarshalJSON([]byte(tt.jsonData), &json)
+			require.Nil(t, err, "failed to unmarshal data JSON")
+
+			ctxIn := context.Background()
+			store := inmem.NewFromObject(json)
+			ctx := context.Background()
+
+			pe, err := policy.NewEngineFromRego(ctx, tt.rego, store)
+			require.Nil(t, err, "failed to create policy engine")
+
+			res, err := pe.Eval(ctxIn, tt.input)
```

Comment:
Looks like the test cases cover only happy paths when executing `Eval`. I think that it would be good to have coverage of failures also.

Reply:
Yup - the way its set up, it's designed so that once a policy engine is successfully created, it should by really hard for `Eval` to actually return an error (I tried a lot of things, including divide by zeros....).. it seems like mainly err returned would be because of a system failure.. This is because of the `vallidatePolicy` function catches a lot of the errors that can happen during policy instantiation, so i've added the case where Eval would fail in the instantiation tests here instead:

https://github.com/lumjjb/spire/blob/policy_opa/pkg/server/authpolicy/policy_test.go#L418-L427